boba/cli.py

```diff
@@ -64,19 +100,65 @@ def run(folder, run_all, num, thru):
 
     Run a range of universes for example 1 through 5: boba run 1 --thru 5
     """
-
+    
     check_path(folder)
 
-    if num < 0 and not run_all:
-        print_help('Error: Missing argument "NUM".')
-
-    cmd = ['sh', 'execute.sh']
-    if not run_all:
-        cmd.append(str(num))
-        if thru > 0:
-            cmd.append(str(thru))
-
-    subprocess.run(cmd, cwd=folder)
+    # get the names of all the universes we want to run
+    universes = []
+    data = pd.read_csv(folder + "/summary.csv")
+    vals = data['Filename'].values
+    file_extension = vals[0].split('.')[1]
+    if run_all:
+        for universe in vals:
+            universes.append(universe)
+    else:
+        if thru == -1:
+            thru = num
+
+        for i in range(num, thru + 1):
+            universe = "universe_" + str(i) + "." + file_extension
+            universes.append(universe)
+
+    run_commands_in_folder(folder, "pre_exe")
+
+    if jobs == 0:
+        jobs = mp.cpu_count()
+
+    if batch_size == 0:
+        batch_size = min(int(len(universes)**(0.5)), int(len(universes)/jobs) + 1)
+
+    pool = mp.Pool(jobs)
+
+    results = []
+
+    # callback that is run for each retrieved result.
+    def check_result(r):
+        results.extend(r)
+        for res in r:
```

Comment:
Perhaps you can save the current results list to disk here, to avoid losing data if the program terminates prematurely